diff --git a/Android.mk b/Android.mk
index 1051b448740f32cfb65f6c4ad82e7e5904f524cf..815ce507635c5ab20210c11fda75d82a04d0515a 100644
--- a/Android.mk
+++ b/Android.mk
@@ -50,0 +51 @@ LOCAL_CFLAGS += -DFD_SET_DEFINED -DSOCKLEN_T_DEFINED -DENABLE_LIBRARY
+LOCAL_CFLAGS += -DPKGNAME=com/tcptun/client -DCLSNAME=HevSocks5Tunnel
diff --git a/src/hev-jni.c b/src/hev-jni.c
index 96985555ca9a82b00d548d0efd24bf6ae456fe19..94b83eca1340f4fccf6e4c0b6a601e967c91fefc 100644
--- a/src/hev-jni.c
+++ b/src/hev-jni.c
@@ -45 +45,2 @@ struct _ThreadData
-static int is_working;
+static volatile int is_working;
+static int thread_created;
@@ -53,0 +55 @@ static void native_stop_service (JNIEnv *env, jobject thiz);
+static jboolean native_is_running (JNIEnv *env, jobject thiz);
@@ -59,0 +62 @@ static JNINativeMethod native_methods[] = {
+    { "TProxyIsRunning", "()Z", (void *)native_is_running },
@@ -97,0 +101,2 @@ thread_handler (void *data)
+    is_working = 0;
+
@@ -112,0 +118,5 @@ native_start_service (JNIEnv *env, jobject thiz, jstring config_path, jint fd)
+    if (thread_created && !is_working) {
+        pthread_join (work_thread, NULL);
+        thread_created = 0;
+    }
+
@@ -130,0 +141 @@ native_start_service (JNIEnv *env, jobject thiz, jstring config_path, jint fd)
+    thread_created = 1;
@@ -140 +151 @@ native_stop_service (JNIEnv *env, jobject thiz)
-    if (!is_working)
+    if (!thread_created)
@@ -143 +154,2 @@ native_stop_service (JNIEnv *env, jobject thiz)
-    hev_socks5_tunnel_quit ();
+    if (is_working)
+        hev_socks5_tunnel_quit ();
@@ -146,0 +159 @@ native_stop_service (JNIEnv *env, jobject thiz)
+    thread_created = 0;
@@ -150,0 +164,6 @@ exit:
+static jboolean
+native_is_running (JNIEnv *env, jobject thiz)
+{
+    return is_working ? JNI_TRUE : JNI_FALSE;
+}
+
